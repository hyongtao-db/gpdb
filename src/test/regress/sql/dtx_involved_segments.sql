-- This test is used to test if the segments involved in a dtx is correct

CREATE TABLE test_dispatch_normal (c1 int, c2 int) DISTRIBUTED BY (c1);

-- Case 1: replicated table 

CREATE TABLE test_dispatch_rep (c1 int, c2 int) DISTRIBUTED REPLICATED;

INSERT INTO test_dispatch_rep values (1,1),(2,2),(3,3);
INSERT INTO test_dispatch_normal values (1,2),(2,3),(5,6);


SELECT gp_segment_id, * FROM test_dispatch_normal;

-- Test 3 times to make sure with the same gp_session_id, but update the tuples of
-- different segmets, all the 3 DTX will use 1PC, and dispatch commit message to only
-- 1 segment.
SET Test_print_direct_dispatch_info = ON;
SET optimizer = OFF;

BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
SELECT * FROM test_dispatch_rep;
COMMIT;

BEGIN;
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
SELECT * FROM test_dispatch_rep;
COMMIT;

BEGIN;
UPDATE test_dispatch_normal SET c2 = 5 WHERE c1 = 5;
SELECT * FROM test_dispatch_rep;
COMMIT;

-- Case 2: Normal table
SET Test_print_direct_dispatch_info = ON;
SET optimizer = OFF;

-- 1PC, dispatch commit message to 1 segments 
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
SELECT * FROM test_dispatch_normal WHERE c1 = 1;
COMMIT;

-- 1PC, but dispatch commit message to 2 segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
SELECT * FROM test_dispatch_normal WHERE c1 = 1 or c1 = 2;
COMMIT;

-- 1PC, but dispatch commit message to all segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
SELECT * FROM test_dispatch_normal;
COMMIT;

-- 2PC, dispatch prepare message to 2 segments, dispatch commit prepared message to 2 segments
-- dispatch commit one-phase message to 1 segment
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
SELECT * FROM test_dispatch_normal WHERE c1 = 5; 
COMMIT;

-- 2PC, dispatch prepare and commit message to all segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
UPDATE test_dispatch_normal SET c2 = 5 WHERE c1 = 5;
SELECT * FROM test_dispatch_normal WHERE c1 = 5; 
COMMIT;

-- Clean up
DROP TABLE test_dispatch_rep;
DROP TABLE test_dispatch_normal;