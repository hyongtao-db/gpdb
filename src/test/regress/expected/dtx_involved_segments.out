-- This test is used to test if the segments involved in a dtx is correct
CREATE TABLE test_dispatch_normal (c1 int, c2 int) DISTRIBUTED BY (c1);
-- Case 1: replicated table 
CREATE TABLE test_dispatch_rep (c1 int, c2 int) DISTRIBUTED REPLICATED;
INSERT INTO test_dispatch_rep values (1,1),(2,2),(3,3);
INSERT INTO test_dispatch_normal values (1,2),(2,3),(5,6);
SELECT gp_segment_id, * FROM test_dispatch_normal;
 gp_segment_id | c1 | c2 
---------------+----+----
             1 |  1 |  2
             2 |  5 |  6
             0 |  2 |  3
(3 rows)

-- Test 3 times to make sure with the same gp_session_id, but update the tuples of
-- different segmets, all the 3 DTX will use 1PC, and dispatch commit message to only
-- 1 segment.
SET Test_print_direct_dispatch_info = ON;
SET optimizer = OFF;
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_rep;
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to SINGLE content
BEGIN;
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_rep;
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to SINGLE content
BEGIN;
UPDATE test_dispatch_normal SET c2 = 5 WHERE c1 = 5;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_rep;
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  1 |  1
  2 |  2
  3 |  3
(3 rows)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to SINGLE content
-- Case 2: Normal table
SET Test_print_direct_dispatch_info = ON;
SET optimizer = OFF;
-- 1PC, dispatch commit message to 1 segments 
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_normal WHERE c1 = 1;
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  1 |  1
(1 row)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to SINGLE content
-- 1PC, but dispatch commit message to 2 segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_normal WHERE c1 = 1 or c1 = 2;
INFO:  (slice 1) Dispatch command to PARTIAL contents: 1 0
 c1 | c2 
----+----
  1 |  1
  2 |  2
(2 rows)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to PARTIAL contents: 0 1
-- 1PC, but dispatch commit message to all segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_normal;
INFO:  (slice 1) Dispatch command to ALL contents: 0 1 2
 c1 | c2 
----+----
  2 |  2
  5 |  5
  1 |  1
(3 rows)

COMMIT;
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to ALL contents: 0 1 2
-- 2PC, dispatch prepare message to 2 segments, dispatch commit prepared message to 2 segments
-- dispatch commit one-phase message to 1 segment
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_normal WHERE c1 = 5; 
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  5 |  5
(1 row)

COMMIT;
INFO:  Distributed transaction has PARTIAL contents: 0 1  persisted WAL
INFO:  Distributed transaction command 'Distributed Prepare' to PARTIAL contents: 0 1
INFO:  Distributed transaction has PARTIAL contents: 0 1  persisted WAL
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2
-- 2PC, dispatch prepare and commit message to all segments
BEGIN;
UPDATE test_dispatch_normal SET c2 = 1 WHERE c1 = 1;
INFO:  (slice 0) Dispatch command to SINGLE content
UPDATE test_dispatch_normal SET c2 = 2 WHERE c1 = 2;
INFO:  (slice 0) Dispatch command to SINGLE content
UPDATE test_dispatch_normal SET c2 = 5 WHERE c1 = 5;
INFO:  (slice 0) Dispatch command to SINGLE content
SELECT * FROM test_dispatch_normal WHERE c1 = 5; 
INFO:  (slice 1) Dispatch command to SINGLE content
 c1 | c2 
----+----
  5 |  5
(1 row)

COMMIT;
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Prepare' to ALL contents: 0 1 2
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2
-- Clean up
DROP TABLE test_dispatch_rep;
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Prepare' to ALL contents: 0 1 2
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2
DROP TABLE test_dispatch_normal;
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Prepare' to ALL contents: 0 1 2
INFO:  Distributed transaction has ALL contents: 0 1 2  persisted WAL
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2